File: Events/2018/05/05.rs
aoc::main!();

fn preprocessing(input_: &str) -> String {
    input_.to_string()
}


fn part_1(polymer: String) -> usize {
    react(polymer)
}


fn part_2(polymer: String) -> usize {
    zip('a'..='z', 'A'..='Z')
    .map(|(l, r)| polymer.clone().replace(r, "").replace(l, ""))
    .map(|poly| react(poly))
    .min()
    .unwrap()
}


fn react(polymer: String) -> usize {
    successors(Some(polymer.clone()), |polymer_|
        Some(
            zip('a'..='z', 'A'..='Z')
            .map(|(a, b)| 
                (
                    [a.to_string(), b.to_string()].concat()
                ,
                    [b.to_string(), a.to_string()].concat()
                    ))
        .fold(polymer_.to_string(), |poly: String, (l, r)|
            poly.replace(&r, "").replace(&l, ""))))
    .tuple_windows()
    .take_while(|(p1, p2)| p1 != p2)
    .last()
    .unwrap()
    .1
    .len()
}

File: events/2018/05/05.py
import string 

def preprocessing(puzzle_input):
    return puzzle_input

def solver(polymer_):
    units            = list(zip(string.ascii_lowercase, string.ascii_uppercase))
    new_length       = 0 
    shortest_polymer = len(polymer_)
    
    for unit_low, unit_up in [('', '')] + units:
        polymer = polymer_.replace(unit_low, '').replace(unit_up, '')
        
        length = len(polymer)
        while length != new_length:
            length = len(polymer)
            for l, u in units:
                polymer = polymer.replace(l + u,'').replace(u + l,'')
            new_length = len(polymer)
        shortest_polymer = min(shortest_polymer, new_length)
        
        if unit_low == '' : yield shortest_polymer
    
    yield shortest_polymer